docs/geps/gep-05.rst

```diff
@@ -0,0 +1,198 @@
+
+
+Motivation and Scope
+--------------------
+
+For several taxes and transfers, German law specifies that these be rounded in specific
+ways. This leads to different use cases.
+
+1. Some applications require the exact, rounded, amount as specified in the law. This
+   is also helpful for creating test cases.
+2. Other applications require smooth functions and the non-rounding error does not
+   matter much.
+
+This document describes how we go about supporting both use cases.
+
+
+Implementation
+--------------
+
+GETTSIM allows for optional rounding of functions' results. Rounding parameters
+are specified in the ``.yaml``-files.
+
+The law on the public pension insurance specifies that the maximum possible
+Grundrentenzuschlag ``höchstwert_grundr_zuschlag_m`` be rounded to the nearest
+fourth decimal point (§76g SGB VI: Zuschlag an Entgeltpunkten für langjährige
+Versicherung). The example below contains GETTSIM's encoding of this fact.
+
+The snippet is taken from ``ges_rentenv.yaml``, which contains the following
+code:
+
+.. code-block:: yaml
+
+    rounding:
+      höchstwert_grundr_zuschlag_m:
+        2020-01-01:
+          base: 0.0001
+          direction: nearest
+          reference: §76g SGB VI Abs. 4 Nr. 4
+
+The specification of the rounding parameters starts with the key ``rounding`` at
+the outermost level of indentation. The keys are names of functions.
+
+At the next level, the ``YYYY-MM-DD`` key(s) indicate when rounding was
+introduced and/or changed. This is done in in the same way as for other policy
+parameters as described in :ref:`gep-3`. Those ``YYYY-MM-DD`` key(s) are
+associated with a dictionary of the following elements:
+
+- The parameter ``base`` determines the base to which the variables is rounded.
+  It has to be a floating point number.
+- The parameter ``direction`` has to be one of ``up``, ``down``, or ``nearest``.
+- The reference must contain the reference to the law, which specifies the
+  rounding.
+
+In the same way as other policy parameters, the rounding parameters become part
+of the dictionary ``policy_params``.
+
+A function to be rounded must be decorated with ``add_rounding_spec``. This
+decorator indicates that the output should be potentially rounded.
+``add_rounding_spec`` takes one required argument ``func_arg``, which specifies
+the (parameters) argument to the function containing the rounding as a string.
+In the above example, the rounding specification
+``höchstwert_grundr_zuschlag_m`` will be found in its argument
+``ges_rentenv_params``, Hence the ``func_arg`` argument to ``add_rounding_spec``
+has to be ``"ges_rentenv_params"``:
+
+.. code-block:: python
+
+    @add_rounding_spec(func_arg="ges_rentenv_params")
+    def höchstwert_grundr_zuschlag_m(
+        grundrentenzeiten: IntSeries, ges_rentenv_params: dict
+    ) -> FloatSeries:
+        ...
+        return out
+
+The decorator adds the attribute ``__rounding_arg__`` to the function. When
```

Comment:
Great change to use the name of the parameter dict instead of the file!

But can you explain the rational behind the name func_arg? Why is it different than the attribute later?

Reply:
My thinking was 
- `func_arg` as an argument to the decorator to make clear the name has to come from the function, `arg` is just too generic.
- for `__rounding_arg__`, I thought it was clearer without the `func`. I changed that from `__rounding_parameter_file__`, but it probably should have been `__rounding_spec__` in the first place? That is, it will contain base/direction, as in the (otherwise outdated) implementation in #316?

Reply:
> func_arg as an argument to the decorator to make clear the name has to come from the function

I don't understand. To my understanding, the name comes from the name of the parameter dictionary.

> That is, it will contain base/direction, as in the (otherwise outdated) implementation in implement decorator for rounding. #316?

No, I won't contain the parameters as those needed to be read (including the correct date) from the parameter file when the function is read.

Reply:
So I would prefer ``__rounding_parameter_file__`` for both the attribute name and the argument of the decorator.

Reply:
> No, I won't contain the parameters as those needed to be read (including the correct date) from the parameter file when the function is read.

I do not understand this sentence.

Reply:
> So I would prefer `__rounding_parameter_file__` for both the attribute name and the argument of the decorator.

My point is that with what I am suggesting, you only need to look at this code snippet to understand what is going on.

```python
    @add_rounding_spec(func_arg="ges_rentenv_params")
    def höchstwert_grundr_zuschlag_m(
        grundrentenzeiten: IntSeries, ges_rentenv_params: dict
    )
    pass
```

The rounding specification will be in the dictionary `ges_rentenv_params`, which is an argument to the function. The fact that it comes from some parameters file is irrelevant at this point and downward misleading for user-provided functions / ex post changes to the parameters dictionary.